File: dotnetcore/Dependencies/CustomDependencyBulk/CustomDependencyBulkImpl/ProductDependency.cs
// ===============================================================================
// Alachisoft (R) NCache Sample Code.
// ===============================================================================
// Copyright © Alachisoft.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
// ===============================================================================

using System;
using System.Collections.Generic;
using System.Data.OleDb;
using System.Data.SqlClient;
using System.Globalization;
using System.Linq;
using Alachisoft.NCache.Runtime.Dependencies;

namespace Alachisoft.NCache.Samples
{
    /// <summary>
    /// Implementation of BulkExtensibleDependency.
    /// </summary>
    [Serializable]
    public class ProductDependency : BulkExtensibleDependency
    { 
        private int productId;
        private int unitsInStock;
        private string _connString;

        /// <summary>
        /// Custom constructor initialize dependency
        /// </summary>
        public ProductDependency(string connString, int Id, int units)
        {
            _connString = connString;
            productId = Id;
            unitsInStock = units;
        }

        /// <summary>
        /// Initialize custom dependency
        /// </summary>
        public override bool Initialize()
        { 
            return true;
        }

        ///<summary>
        /// Get units in stock of each product
        /// </summary>
        private Dictionary<int,int> UnitsInStockStatus(List<int> productIds)
        {
            using (var connection = new SqlConnection(_connString))
            {

                var ids =
                productIds.Select(i => i.ToString(CultureInfo.InvariantCulture))
                    .Aggregate((s1, s2) => s1 + ", " + s2);

                string queryString =
                    "select ProductID, UnitsInStock " +
                    "from Products " +
                    $"WHERE ProductID in ({ids})";

                Dictionary<int, int> productsInfo = new Dictionary<int, int>();

                using (var cmd = new SqlCommand(queryString, connection))
                {
                    connection.Open();

                    using (SqlDataReader reader = cmd.ExecuteReader())
                    {

                        while (reader.Read())
                        {
                            int units = 0;
                            Int32.TryParse(reader["UnitsInStock"].ToString(), out units);
                            productsInfo.Add((int)reader.GetValue(0), units);
                        }
                    }
                }

                return productsInfo;
            }
        }

        /// <summary>
        /// This method is used to invalidate a cached items on clean interval
        /// </summary> 
        public override void EvaluateBulk(IEnumerable<BulkExtensibleDependency> dependencies)
        {
            List<ProductDependency> dependencyList = new List<ProductDependency>();
            foreach (ProductDependency productDependency in dependencies)
            {
                dependencyList.Add(productDependency);
            }

            List<int> productIds = GetDependencyProductIds(dependencyList);
            var productsInfo = UnitsInStockStatus(productIds);

            foreach (ProductDependency productDependency in dependencyList)
            {
                if (productsInfo.ContainsKey(productDependency.productId))
                {
                    if (productsInfo[productDependency.productId] != productDependency.unitsInStock)
                    {
                        productDependency.Expire();
                    }
                }
            }
        }

        /// <summary>
        /// Get list of productID of Products with bulk dependency
        /// </summary>
        /// <param name="dependencies">IEnumerable BulkExtensibleDependency collection</param>
        /// <returns>List of productId Products with bulk dependency</returns>
        private List<int> GetDependencyProductIds(IEnumerable<BulkExtensibleDependency> dependencies)
        {
            List<int> productIds = new List<int>();

            foreach(ProductDependency productDependency in dependencies)
            {
                productIds.Add(productDependency.productId);
            }
            
            return productIds;
        }

        /// <summary>
        /// Dispose custom dependency
        /// </summary>

        protected override void DependencyDispose()
        {
            base.DependencyDispose();
        }

        public override string ToString()
        {
            return$"Product ID: {productId} UnitsinStock: {unitsInStock}";
        }

    }

}

File: java/BulkCustomDependency/src/main/java/com/Alachisoft/NCache/Samples/BulkCustomDependencyImpl/ProductDependency.java
package com.Alachisoft.NCache.Samples.BulkCustomDependencyImpl;

import com.alachisoft.ncache.runtime.dependencies.BulkExtensibleDependency;

import java.io.Serializable;
import java.sql.*;
import java.util.*;

/**
 * Implementation of BulkExtensibleDependency.
 */
public class ProductDependency extends BulkExtensibleDependency implements Serializable {

    private int productId;
    private int unitsInStock;
    private final String _connString;
    private final String _userId;
    private final String _password;

    private transient  Connection _conn;

    /**
     * Custom constructor initialize dependency
     * @param connString
     * @param id
     * @param units
     * @param userId
     * @param password
     */
    public ProductDependency(String connString, String userId, String password, int id, int units) {
        _connString = connString;
        _userId = userId;
        _password = password;
        unitsInStock = units;
        productId = id;
    }

    /**
     * This method is used to invalidate a cached items on clean interval
     * @param iterable
     * @throws Exception
     */
    @Override
    public void evaluateBulk(Iterable<BulkExtensibleDependency> iterable) throws Exception {
        List<BulkExtensibleDependency> dependencyList = new ArrayList<>();
        for (BulkExtensibleDependency dep : iterable) {
            dependencyList.add(dep);
        }

        List<Integer> productIds = getDependencyProductIds(dependencyList);

        Map<Integer, Integer> productInfo = unitsInStockStatus(productIds);

        for (BulkExtensibleDependency dep : dependencyList) {
            ProductDependency productDependency = (ProductDependency)dep;
            if(productInfo.containsKey(productDependency.productId))
            {
                if(productInfo.get(productDependency.productId) != productDependency.unitsInStock)
                {
                    dep.expire();
                }
            }
        }
    }

    /**
     * Get list of productID of Products with bulk dependency
     * @param dependencyList IEnumerable BulkExtensibleDependency collection
     * @return List of productId Products with bulk dependency
     */
    private List<Integer> getDependencyProductIds(List<BulkExtensibleDependency> dependencyList) {
        List<Integer> productIds = new ArrayList<>();

        for (BulkExtensibleDependency dependency:
             dependencyList) {
            ProductDependency productDependency = (ProductDependency)dependency;
            productIds.add(productDependency.productId);
        }
        return productIds;
    }

    /**
     *  Initialize custom dependency
     * @return
     */
    @Override
    public boolean initialize() {
        return true;
    }

    /**
     * Establishes the database connection
     * @param connString
     * @param user
     * @param pass
     */
    private void establishConnection(String connString, String user, String pass){
        try {
            if(connString == null)
                throw new IllegalArgumentException();
            if(user == null)
                throw new IllegalArgumentException();
            if(pass == null)
                throw new IllegalArgumentException();
            _conn = DriverManager.getConnection(connString, user, pass);
        } catch (SQLException throwable) {
            throwable.printStackTrace();
        }
    }

    /**
     * Get units in stock of each product
     * @param productIds
     * @return
     * @throws SQLException
     */
    private Map<Integer, Integer> unitsInStockStatus(List<Integer> productIds) throws SQLException {
        Map<Integer, Integer> productInfo = new HashMap<>();
        establishConnection(_connString, _userId, _password);
        String query = "Select PRODUCTID, UNITSINSTOCK from SCOTT.PRODUCTS WHERE PRODUCTID = ?";
        ResultSet rs = null;
        for(int id : productIds){
            PreparedStatement ps = _conn.prepareStatement(query);
            ps.setInt(1, id);
            rs = ps.executeQuery();
            while (rs.next()){
                productId = rs.getInt("PRODUCTID");
                unitsInStock = rs.getInt("UNITSINSTOCK");
                productInfo.put(productId, unitsInStock);
            }
            rs.close();
        }
        return productInfo;
    }

    /**
     * Dispose custom dependency
     */
    @Override
    public void dispose() {
        if(_conn != null) {
            try {
                _conn.close();
            } catch (SQLException throwables) {
                throwables.printStackTrace();
            }
        }
        super.dispose();
    }

    @Override
    public String toString() {
        return "ProductDependency{" +
                "productId=" + productId +
                ", unitsInStock=" + unitsInStock +
                '}';
    }
}
